--- doc/sphinx/Makefile.in.orig	2015-10-04 09:30:26.260411617 +0200
+++ doc/sphinx/Makefile.in	2015-10-04 09:31:05.661090770 +0200
@@ -602,6 +602,7 @@
 	rm -rf $(BUILDDIR)
 
 include/params.rst: $(top_builddir)/bin/varnishd/varnishd
+	LD_LIBRARY_PATH=$(top_builddir)/lib/libvarnish/.libs:$(top_builddir)/lib/libvarnishcompat/.libs:$(top_builddir)/lib/libvcc/.libs:$(top_builddir)/lib/libvgz/.libs \
 	$(top_builddir)/bin/varnishd/varnishd -x dumprstparam > include/params.rst
 
 # XXX add varnishstat here when it's been _opt2rst'ed
