--- doc/sphinx/Makefile.in.orig	2014-10-08 09:48:47.000000000 +0200
+++ doc/sphinx/Makefile.in	2014-11-25 22:49:55.137641462 +0100
@@ -517,6 +517,7 @@
 	rm -rf $(BUILDDIR)
 
 include/params.rst: $(top_builddir)/bin/varnishd/varnishd
+	LD_LIBRARY_PATH=$(top_builddir)/lib/libvarnish/.libs:$(top_builddir)/lib/libvarnishcompat/.libs:$(top_builddir)/lib/libvcc/.libs:$(top_builddir)/lib/libvgz/.libs \
 	$(top_builddir)/bin/varnishd/varnishd -x dumprstparam > include/params.rst
 
 # XXX add varnishstat here when it's been _opt2rst'ed
