diff -Naur ../varnish-2.1.4/redhat/varnish.initrc ./redhat/varnish.initrc
--- ../varnish-2.1.4/redhat/varnish.initrc	2010-10-21 10:57:22.000000000 +0200
+++ ./redhat/varnish.initrc	2010-11-03 15:20:07.663331341 +0100
@@ -12,6 +12,8 @@
 # Provides: varnish
 # Required-Start: $network $local_fs $remote_fs
 # Required-Stop: $network $local_fs $remote_fs
+# Default-Start:
+# Default-Stop:
 # Should-Start: $syslog
 # Short-Description: start and stop varnishd
 # Description: Varnish is a high-perfomance HTTP accelerator
diff -Naur ../varnish-2.1.4/redhat/varnishlog.initrc ./redhat/varnishlog.initrc
--- ../varnish-2.1.4/redhat/varnishlog.initrc	2010-10-21 10:57:22.000000000 +0200
+++ ./redhat/varnishlog.initrc	2010-11-03 15:20:07.664330786 +0100
@@ -12,6 +12,8 @@
 # Provides: varnishlog
 # Required-Start: $network $local_fs $remote_fs
 # Required-Stop: $network $local_fs $remote_fs
+# Default-Start:
+# Default-Stop:
 # Short-Description: start and stop varnishlog
 # Description: Varnish HTTP accelerator logging daemon
 ### END INIT INFO
diff -Naur ../varnish-2.1.4/redhat/varnishncsa.initrc ./redhat/varnishncsa.initrc
--- ../varnish-2.1.4/redhat/varnishncsa.initrc	2010-10-21 10:57:22.000000000 +0200
+++ ./redhat/varnishncsa.initrc	2010-11-03 15:20:07.664330786 +0100
@@ -12,6 +12,8 @@
 # Provides: varnishncsa
 # Required-Start: $network $local_fs $remote_fs
 # Required-Stop: $network $local_fs $remote_fs
+# Default-Start:
+# Default-Stop:
 # Short-Description: start and stop varnishncsa
 # Description: Varnish HTTP accelerator logging daemon
 ### END INIT INFO
