Index: bin/varnishtest/vtc_server.c
===================================================================
--- bin/varnishtest/vtc_server.c	(revision 3155)
+++ bin/varnishtest/vtc_server.c	(revision 3157)
@@ -97,7 +97,7 @@
 		vtc_log(vl, 3, "Accepted socket fd is %d", fd);
 		http_process(vl, s->spec, fd, 0);
 		vtc_log(vl, 3, "shutting fd %d", fd);
-		AZ(shutdown(fd, SHUT_WR));
+		assert((shutdown(fd, SHUT_WR) == 0) || errno == ENOTCONN);
 		TCP_close(&fd);
 	}
 	vtc_log(vl, 2, "Ending");
