--- bin/varnishtest/tests/v00017.vtc.orig	2015-09-30 13:39:14.000000000 +0200
+++ bin/varnishtest/tests/v00017.vtc	2015-10-05 13:58:05.129191488 +0200
@@ -36,9 +36,9 @@
 	sub vcl_recv { if (client.ip ~ a) { return(pass); } }
 }
 
-varnish v1 -errvcl {DNS lookup(10.1.2.): } {
+varnish v1 -errvcl {DNS lookup(198..2.): } {
 	backend b { .host = "127.0.0.1"; }
-	acl a { "10.1.2."; }
+	acl a { "198..2."; }
 	sub vcl_recv { if (client.ip ~ a) { return(pass); } }
 }
 
