--- doc/sphinx/Makefile.in.orig	2016-01-29 08:47:23.963344825 +0100
+++ doc/sphinx/Makefile.in	2016-01-29 08:49:30.290601970 +0100
@@ -531,6 +531,7 @@
 	rm -rf $(BUILDDIR)
 
 include/params.rst: $(top_builddir)/bin/varnishd/varnishd
+	LD_LIBRARY_PATH=$(top_builddir)/lib/libvarnish/.libs:$(top_builddir)/lib/libvarnishcompat/.libs:$(top_builddir)/lib/libvcc/.libs:$(top_builddir)/lib/libvgz/.libs \
 	$(top_builddir)/bin/varnishd/varnishd -x dumprstparam > include/params.rst
 
 include/counters.rst: $(top_builddir)/bin/varnishstat/vsc2rst
