diff -Naur ../varnish-2.0.2.orig/bin/varnishtest/Makefile.am ./bin/varnishtest/Makefile.am
--- ../varnish-2.0.2.orig/bin/varnishtest/Makefile.am	2008-11-14 13:24:08.000000000 +0100
+++ ./bin/varnishtest/Makefile.am	2008-11-17 13:40:05.964237951 +0100
@@ -1,6 +1,6 @@
 # $Id: Makefile.am 3374 2008-11-10 10:12:28Z tfheen $
 
-TESTS_ENVIRONMENT = ./varnishtest
+TESTS_ENVIRONMENT = ./varnishtest -v
 TESTS = $(srcdir)/tests/*.vtc
 
 DISTCLEANFILES = _.ok
diff -Naur ../varnish-2.0.2.orig/bin/varnishtest/Makefile.in ./bin/varnishtest/Makefile.in
--- ../varnish-2.0.2.orig/bin/varnishtest/Makefile.in	2008-11-14 13:59:33.000000000 +0100
+++ ./bin/varnishtest/Makefile.in	2008-11-17 13:40:51.503384938 +0100
@@ -204,7 +204,7 @@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 varnishconfdir = @varnishconfdir@
-TESTS_ENVIRONMENT = ./varnishtest
+TESTS_ENVIRONMENT = ./varnishtest -v
 TESTS = $(srcdir)/tests/*.vtc
 DISTCLEANFILES = _.ok
 INCLUDES = -I$(top_srcdir)/include
