diff -Naur ../varnish-2.0-beta1.orig/redhat/README.redhat ./redhat/README.redhat
--- ../varnish-2.0-beta1.orig/redhat/README.redhat	2008-08-27 09:45:40.000000000 +0200
+++ ./redhat/README.redhat	2008-09-02 16:14:43.000000000 +0200
@@ -5,6 +5,24 @@
 
 Varnish should work fine with GCC 3.3 and above.
 
+Upgrading from 1.x to 2.0
+=========================
+There are a few changes in the vcl language from varnish-1.x to 2.0.
+Because of varnish' dynamic vcl loading feature, there is no way to
+guarantee that the vcl file in use actually exists on disk. Thus,
+there is no way to securely automate this process, and one must do the
+changes by hand.
+
+In vcl, the word "insert" has been replaced by "deliver". 
+
+In the vcl declaration of backends, where one earlier used "set
+backend", backend parts are now just prefixed with a dot, so the
+default localhost configuration will look like this:
+
+backend default {
+  .host = "127.0.0.1";
+  .port = "80";
+}
 
 Configuration of addresses and ports
 ====================================
