--- ../varnish-2.1.orig/man/vcl.7so
+++ man/vcl.7so
@@ -126,7 +126,6 @@ it's saintmode list
 can be set to the maximum list size. Setting a value of 0 disables
 saintmode checking entirely for that backend. The value in the backend
 declaration overrides the parameter.
-
 .Ss Directors
 Directors choose from different backends based on health status and a
 per-director algorithm.
--- ../varnish-2.1.orig/bin/varnishd/varnishd.1
+++ bin/varnishd/varnishd.1
@@ -173,6 +173,9 @@ to the specified
 See
 .Sx Run-Time Parameters
 for a list of parameters.
+.It Fl S Ar file
+Path to a file containing a secret used for authorizing access to the
+management port.
 .It Fl s Ar type Ns Xo
 .Op , Ns Ar options
 .Xc
--- ../varnish-2.1.orig/redhat/varnish.sysconfig
+++ ./redhat/varnish.sysconfig
@@ -38,6 +38,7 @@
              -T localhost:6082 \
              -f /etc/varnish/default.vcl \
              -u varnish -g varnish \
+             -S /etc/varnish/secret \
              -s file,/var/lib/varnish/varnish_storage.bin,1G"
 
 
