diff -Naur ../varnish-2.1.2.orig/lib/libjemalloc/jemalloc_linux.c ./lib/libjemalloc/jemalloc_linux.c
--- ../varnish-2.1.2.orig/lib/libjemalloc/jemalloc_linux.c	2010-05-05 09:32:02.000000000 +0200
+++ ./lib/libjemalloc/jemalloc_linux.c	2010-07-29 15:39:00.221232248 +0200
@@ -273,6 +273,11 @@
 #  define QUANTUM_2POW		4
 #  define SIZEOF_PTR_2POW	2
 #endif
+#ifdef __s390__
+#  define PAGESIZE_2POW		12
+#  define QUANTUM_2POW		4
+#  define SIZEOF_PTR_2POW	2
+#endif
 #ifdef __s390x__
 #  define PAGESIZE_2POW		12
 #  define QUANTUM_2POW		4
