diff -Naur ../varnish-2.0-beta2.orig/lib/libjemalloc/jemalloc_linux.c ./lib/libjemalloc/jemalloc_linux.c
--- ../varnish-2.0-beta2.orig/lib/libjemalloc/jemalloc_linux.c	2008-09-24 20:05:19.000000000 +0200
+++ ./lib/libjemalloc/jemalloc_linux.c	2008-10-06 14:13:28.950350627 +0200
@@ -257,7 +257,7 @@
 #  define NO_TLS
 #endif
 #ifdef __powerpc__
-#  define PAGESIZE_2POW		12
+#  define PAGESIZE_2POW		16
 #  define QUANTUM_2POW		4
 #  define SIZEOF_PTR_2POW	2
 #endif
diff -Naur ../varnish-2.0-beta2.orig/lib/libjemalloc/malloc.c ./lib/libjemalloc/malloc.c
--- ../varnish-2.0-beta2.orig/lib/libjemalloc/malloc.c	2008-09-24 20:05:19.000000000 +0200
+++ ./lib/libjemalloc/malloc.c	2008-10-06 14:13:42.023005090 +0200
@@ -261,7 +261,7 @@
 #  define NO_TLS
 #endif
 #ifdef __powerpc__
-#  define PAGESIZE_2POW		12
+#  define PAGESIZE_2POW		16
 #  define QUANTUM_2POW		4
 #  define SIZEOF_PTR_2POW	2
 #endif
